File: FinalProject/2_GraphAndTree/Baitoan4/FromAdjMat.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n;
    cin >> n;
    int type; cin >> type; // 0: undirected, 1: directed
    bool directed = (type == 1);

    vector<vector<int>> adjMatrix(n, vector<int>(n));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> adjMatrix[i][j];

    vector<tuple<int, int, int>> edges;
    vector<vector<pair<int, int>>> adj(n);
    vector<unordered_map<int, int>> ogMap(n), icMap(n);

    for (int u = 0; u < n; ++u) {
        for (int v = 0; v < n; ++v) {
            int k = adjMatrix[u][v];
            if (k == 0) continue;

            if (directed) {
                edges.emplace_back(u, v, k);
                adj[u].push_back({v, k});
                ogMap[u][v] += k;
                icMap[v][u] += k;
            } else if (!directed && u <= v) {
                edges.emplace_back(u, v, k);
                if (u == v) {
                    // canh khuyen
                    adj[u].push_back({v, k});
                    ogMap[u][v] += k;
                    icMap[u][v] += k;
                } else {
                    // canh thuong
                    adj[u].push_back({v, k});
                    adj[v].push_back({u, k});
                    ogMap[u][v] += k;
                    ogMap[v][u] += k;
                    icMap[v][u] += k;
                    icMap[u][v] += k;
                }
            }
        }
    }

    // convert unordered_map to vector<pair> for printing
    vector<vector<pair<int, int>>> og(n), ic(n);
    for (int u = 0; u < n; ++u) {
        for (auto& [v, k] : ogMap[u])
            og[u].push_back({v, k});
        for (auto& [v, k] : icMap[u])
            ic[u].push_back({v, k});
    }

    cout << "\nEdge List:\n";
    for (auto [u, v, k] : edges)
        cout << u << " " << v << " " << k << "\n";

    cout << "\nAdjacency List:\n";
    for (int u = 0; u < n; ++u) {
        cout << u << ": ";
        for (auto [v, k] : adj[u])
            cout << "{" << v << "," << k << "} ";
        cout << "\n";
    }

    cout << "\nAdjacency Map (Incoming - ic):\n";
    for (int u = 0; u < n; ++u) {
        cout << u << ": ";
        for (auto [v, k] : ic[u])
            cout << "{" << v << "," << k << "} ";
        cout << "\n";
    }

    cout << "\nAdjacency Map (Outgoing - og):\n";
    for (int u = 0; u < n; ++u) {
        cout << u << ": ";
        for (auto [v, k] : og[u])
            cout << "{" << v << "," << k << "} ";
        cout << "\n";
    }

    return 0;
}

/*
Input 1: Directed graph
6 1
2 3 0 1 0 0
0 0 4 0 0 1
1 0 2 3 0 0
0 1 0 0 5 2
3 0 0 0 1 0
0 0 2 1 0 3

Input 2: Undirected graph
6 0
2 3 1 0 2 0
3 1 2 4 0 2
1 2 3 1 2 0
0 4 1 2 5 1
2 0 2 5 2 3
0 2 0 1 3 4
*/

File: FinalProject/2_GraphAndTree/Baitoan4/FromAdjMat.py
from collections import defaultdict

n, type_ = map(int, input().split())
directed = (type_ == 1)

adjMatrix = [list(map(int, input().split())) for _ in range(n)]

edges = []
adj = [[] for _ in range(n)]
ogMap = [defaultdict(int) for _ in range(n)]
icMap = [defaultdict(int) for _ in range(n)]

for u in range(n):
    for v in range(n):
        k = adjMatrix[u][v]
        if k == 0:
            continue

        if directed:
            edges.append((u, v, k))
            adj[u].append((v, k))
            ogMap[u][v] += k
            icMap[v][u] += k
        else:
            if u <= v:
                edges.append((u, v, k))
                if u == v:
                    # loop edge
                    adj[u].append((v, k))
                    ogMap[u][v] += k
                    icMap[u][v] += k
                else:
                    # normal undirected edge
                    adj[u].append((v, k))
                    adj[v].append((u, k))
                    ogMap[u][v] += k
                    ogMap[v][u] += k
                    icMap[v][u] += k
                    icMap[u][v] += k

og = [[] for _ in range(n)]
ic = [[] for _ in range(n)]

for u in range(n):
    for v, k in ogMap[u].items():
        og[u].append((v, k))
    for v, k in icMap[u].items():
        ic[u].append((v, k))

print("\nEdge List:")
for u, v, k in edges:
    print(u, v, k)

print("\nAdjacency List:")
for u in range(n):
    print(f"{u}: ", end="")
    for v, k in adj[u]:
        print(f"{{{v},{k}}} ", end="")
    print()

print("\nAdjacency Map (Incoming - ic):")
for u in range(n):
    print(f"{u}: ", end="")
    for v, k in ic[u]:
        print(f"{{{v},{k}}} ", end="")
    print()

print("\nAdjacency Map (Outgoing - og):")
for u in range(n):
    print(f"{u}: ", end="")
    for v, k in og[u]:
        print(f"{{{v},{k}}} ", end="")
    print()

# Input 1: Directed graph
# 6 1
# 2 3 0 1 0 0
# 0 0 4 0 0 1
# 1 0 2 3 0 0
# 0 1 0 0 5 2
# 3 0 0 0 1 0
# 0 0 2 1 0 3
#
# Input 2: Undirected graph
# 6 0
# 2 3 1 0 2 0
# 3 1 2 4 0 2
# 1 2 3 1 2 0
# 0 4 1 2 5 1
# 2 0 2 5 2 3
# 0 2 0 1 3 4
